### PR TITLE
[REVIEW] Fix `scan` unit test and upgrade to more appropriate algorithms

### PR DESCRIPTION
**Pull request by codereport**
This PR addresses https://github.com/rapidsai/cudf/issues/4653

1. Most importantly, for the `max` and `min` scan unit tests, the null masks don't actually matter because the corresponding min and max elements have a valid mask - therefore if you had an "identity" function object that just always passed the value through (aka didn't exclude it), the test would still pass.
**Note:** Look at commit "Fix unit tests (make null mask actually matter)" to see the exact fix for the unit test. It was isolated because it is hard to see amongst the other changes.

2. The initialization of the accumulator in the current `std::transform` is wrong. It doesn't take into account that the null mask might render the first element excluded. Fixed this with `numeric_limits`

3. Using a `std::transform` with an accumulator is just a `std::partial_sum`. Using `thrust::inclusive_scan` is appropriate here, but I have a TODO for author (and others to comment).

### CODE REVIEW

**Comment by codereport** [comment]
@karthikeyann I think you were the original author of these unit tests based on git blame. Are you able to comment on the `TODO`:
```
    // TODO: potentially the original author was trying to avoid `thrust::inclusive_scan` for testing
    // if this is the case, I can just replace this with `std::partial_sum` and lambda wrapped `std::min`
```
If this wasn't the case, I will just remove the comment and leave as is.

**Comment by karthikeyann** [comment]
> @karthikeyann I think you were the original author of these unit tests based on git blame. Are you able to comment on the `TODO`:
> 
> ```
>     // TODO: potentially the original author was trying to avoid `thrust::inclusive_scan` for testing
>     // if this is the case, I can just replace this with `std::partial_sum` and lambda wrapped `std::min`
> ```
> 
> If this wasn't the case, I will just remove the comment and leave as is.

The `cudf::experimental::scan` algorithm used `thrust::inclusive_scan` in its implementation. So using the same to test the API seemed not a good practice.  (Also if `thrust::inclusive_scan` had a bug, it will go unnoticed.) If developer is writing the algorithm and also unit tests, at least he/she shouldn't use the same APIs to create expected results to compare. so, I avoided using `thrust::inclusive_scan`.

**Comment by harrism** [comment]
Can you make it more clear in the description what this PR is fixing? Link to an issue?

Regarding using Thrust vs. STL in the unit tests, since the data are quite small in the tests, I highly recommend using STL in unit tests, and renaming the file to `.cpp`. This will help reduce our compile time, which is greatly increased by using Thrust and `.cu`.

**Comment by codereport** [comment]
> Can you make it more clear in the description what this PR is fixing? Link to an issue?

Link to issue: https://github.com/rapidsai/cudf/issues/4653

**Comment by codereport** [comment]
@harrism @karthikeyann Algorithms have been switch to `std` algorithms and I have change the file to a `.cu` :) Please review when you have a chance 

Note: refer to commit `Switch to non-Thrust algorithms (`std::partial_sum` & `std::transform`)` for actual diff.

File: cpp/tests/reductions/scan_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+        exact.begin(),
+        [] (auto const& a, auto const& b) { return std::min(a, b); });
+
+    // string column without nulls
+    cudf::test::strings_column_wrapper col_nonulls(v.begin(), v.end());
+    cudf::test::strings_column_wrapper expected1(exact.begin(), exact.end());
+    this->scan_test(col_nonulls, expected1,
+                    cudf::experimental::make_min_aggregation(), scan_type::INCLUSIVE);
+
+    auto const STRING_MAX = std::string(1, char(127));
```
Comment (karthikeyann) [inline cpp/tests/reductions/scan_tests.cpp]: UTF-8 max string character used in cudf
`__constant__ char max_string_sentinel[5]{"\xF7\xBF\xBF\xBF"};`

https://github.com/rapidsai/cudf/blob/branch-0.13/cpp/include/cudf/detail/utilities/device_operators.cuh#L85

File: cpp/tests/reductions/scan_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+  cudf::experimental::make_sum_aggregation(), scan_type::INCLUSIVE, include_nulls::YES));
+  cudf::test::fixed_width_column_wrapper<TypeParam> expected_col_out2{
+      out_v.begin(), out_v.end(), out_b.begin()};
+  if(do_print) {
+    print_view(expected_col_out2, "expect = ");
+    print_view(col_out->view(),   "result = ");
+  }
+  cudf::test::expect_column_properties_equal(expected_col_out2, col_out->view());
+  cudf::test::expect_columns_equal(expected_col_out2, col_out->view());
+}
```
Comment (karthikeyann) [inline cpp/tests/reductions/scan_tests.cpp]: is an empty newline required at end of file?

Comment (harrism) [reply cpp/tests/reductions/scan_tests.cpp]: https://unix.stackexchange.com/a/18789/131298

```suggestion

}
```

**Pull request merged**